### PR TITLE
Always use varchar(max) or nvarchar(max) for SQL Server string command parameters

### LINKED ISSUE #5975: EF Core 1.0 RTM not filtering properly on fields of type char
### Steps to reproduce

I have created a Github project for this issue
https://github.com/iberodev/EntityFramework5975
### The issue

Suppose I have a Port entity

```
public class Port
{
    public string Code { get; set; }
    public string Name { get; set; }
}

public class PortMap
{
    public PortMap(EntityTypeBuilder<Port> entityBuilder)
    {
        entityBuilder.HasKey(p => p.Code);

        entityBuilder.Property(p => p.Code)
            .HasColumnType("char(5)");

        entityBuilder.Property(p => p.Name)
            .HasMaxLength(65);
    }
}

```

and the field Code has a `char(5)` type.

and the following data:
Code    Name
AUCNB   Canberra
AUMLB   Melbourne
AUSYD   Sydney
NZAKL   Auckland
NZTRG   Tauranga

if I use Entity to retrieve rows based on a filter (eg: the word `NZ` to retrieve only ports whose code starts with `NZ`), the SQL query Entity generates does not return any results from database.

```
public IEnumerable<Port> GetPortsFilteredInDatabase(string filter = "NZ")
{
    var query = _context.Ports.Where(p => p.Code.StartsWith(filter));
    return query.ToList(); //this list will be empty
}

public IEnumerable<Port> GetPortsFilteredInMemory(string filter = "NZ")
{
    var allPorts = _context.Ports.ToList();
    var query = allPorts.Where(p => p.Code.StartsWith(filter));
    return query.ToList(); //this list is correct
}
```

> Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommandBuilderFactory:Information: Executed DbCommand (0ms) [Parameters=[@__filter_0='?' (Size = 5)], CommandType='Text', CommandTimeout='30']
> SELECT [p].[Code], [p].[Name]
> FROM [Ports] AS [p]
> WHERE [p].[Code] LIKE @__filter_0 + '%'

Notice the question mark instead the `NZ` text when generating the SQL statement. Is that normal?

This was working fine in EF Core RC2.

Also notice that if instead CHAR(5), my code column is NVARCHAR(5) everything works well.
### Further technical details

EF Core version: 1.0.0
Operating system: Windows 10
Visual Studio version: VS2015 update 3

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #5975

### CODE REVIEW

File: src/EFCore.Relational/Storage/Internal/RelationalParameterBuilder.cs [hunk 1]
```diff
@@ -96,7 +96,7 @@ public virtual void AddCompositeParameter(string invariantName, Action<IRelation
             Check.NotEmpty(invariantName, nameof(invariantName));
             Check.NotNull(buildAction, nameof(buildAction));
 
-            var innerList = new RelationalParameterBuilder(TypeMapper);
+            var innerList = Create();
```
Comment (smitpatel) [inline src/EFCore.Relational/Storage/Internal/RelationalParameterBuilder.cs]: Since we are here, change variable name to something better?

Comment (AndriySvyryd) [inline src/EFCore.Relational/Storage/Internal/RelationalParameterBuilder.cs]: `providerSpecificName`?

**Comment by smitpatel** [comment]
While the title and referencing issue says to use `nvarchar(max)` why all the test changes are converting to `nvarchar(4000)`. In certain cases, its fine since it would be key. Is it case for all though?

File: src/EFCore.Relational/Storage/Internal/RelationalParameterBuilder.cs (hunk 1 above)
Comment (smitpatel) [inline src/EFCore.Relational/Storage/Internal/RelationalParameterBuilder.cs]: As simple as `var parameterBuilder = CreateParameterBuilder()` would be fine too. (or add relational word too)

**Comment by AndriySvyryd** [comment]
@smitpatel It seems that SQL Client uses varchar(4000) or nvarchar(4000) if the data fits

**Comment by smitpatel** [comment]
Can we add a test for `nvarchar(max)` too?

**Comment by AndriySvyryd** [comment]
Updated

File: test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs [hunk 2]
```diff
@@ -343,7 +344,7 @@ public virtual void Can_insert_and_read_back_all_mapped_data_types()
 @p17='anyone!' (Nullable = false) (Size = 4000)
 @p18='Gumball Rules OK!' (Nullable = false) (Size = 4000)
 @p19='103.3'
-@p20='don't' (Nullable = false) (Size = 4000)
+@p20='" + entity.NvarcharMax + @"' (Nullable = false) (Size = -1)
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs]: Ideally we are not printing out facets which are default. So size should not be printed for `nvarchar(max)`. File a new issue to check if we have a bug.

Comment (AndriySvyryd) [inline test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs]: The default is 0,  and we correctly don't print it if the value is null or empty

Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs]: err... 0 sized string is default. Ok.

**Comment by ErikEJ** [comment]
providers fyi/beware?

**Pull request merged**